nx/lib/nx/defn/grad.ex

```diff
@@ -424,6 +424,24 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
+  defp grad(:concatenate, [tensors, axis], %{shape: ans_shape}, g, cache) do
+    zero_axes = List.duplicate(0, tuple_size(ans_shape))
+    ans_shape_list = Tuple.to_list(ans_shape)
+
+    tensors
+    |> Enum.reduce({0, []}, fn t, {limit, result} ->
+      t_len = elem(t.shape, axis)
+      current_limit = t_len + limit
+      start = List.replace_at(zero_axes, axis, limit)
+      len = List.replace_at(ans_shape_list, axis, t_len)
+
+      result = [{t, Nx.slice(g, start, len)} | result]
+      {current_limit, result}
+    end)
+    |> elem(1)
+    |> grad_pairs(g, cache)
```

Comment:
My take:

```suggestion
    {pairs, _} = 
      Enum.map_reduce(tensors, limit, fn t, limit ->
        t_len = elem(t.shape, axis)
        current_limit = t_len + limit
        start = List.replace_at(zero_axes, axis, limit)
        len = List.replace_at(ans_shape_list, axis, t_len)
        {{t, Nx.slice(g, start, len)}, current_limit}
      end)

    grad_pairs(pairs, g, cache)
```

I am not a big fan to piping to elem, I tend to prefer pattern matching, but piping here would be fine too.

Reply:
Is there an specific reason? I chose to use it here because the output always has the same format, but I can see this being an issue when the reduce can return tagged tuples `{:ok, _}` | `{:error, _}`